pkg/ceph/osd/filesystem.go

```diff
@@ -0,0 +1,240 @@
+		fmt.Sprintf("--conf=%s", mon.GetConfFilePath(config.rootPath, clusterName)),
+		fmt.Sprintf("--osd-data=%s", config.rootPath),
+		fmt.Sprintf("--osd-uuid=%s", config.uuid.String()),
+		fmt.Sprintf("--monmap=%s", monMapTmpPath),
+		fmt.Sprintf("--keyring=%s", getOSDKeyringPath(config.rootPath)),
+	}
+
+	if !isBluestore(config) {
+		options = append(options, fmt.Sprintf("--osd-journal=%s", getOSDJournalPath(config.rootPath)))
+	}
+
+	// create the OSD file system and journal
+	err = context.ProcMan.Run(
+		fmt.Sprintf("mkfs-osd%d", config.id),
+		"ceph-osd",
+		options...)
+
+	if err != nil {
+		return fmt.Errorf("failed osd mkfs for OSD ID %d, UUID %s, dataDir %s: %+v",
+			config.id, config.uuid.String(), config.rootPath, err)
+	}
+
+	// now that the OSD filesystem has been created, back it up so it can be restored/repaired
+	// later on if needed.
+	if err := backupOSFileSystem(config, clusterName); err != nil {
+		return fmt.Errorf("failed to backup OSD filesystem: %+v", err)
+	}
+
+	// update the scheme to indicate the OSD's filesystem has been created
+	if err := markOSDFileSystemCreated(config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// determines if the given OSD's filesystem has previously been created (via osd mkfs).  It may not exist
+// on disk anymore and therefore needs to be repaired, but this determines if it was ever created in the past.
+func isOSDFilesystemCreated(config *osdConfig) bool {
+	if config.partitionScheme == nil {
+		return false
+	}
+
+	return config.partitionScheme.FSCreated
+}
+
+// marks the given OSD's filesystem as created
+func markOSDFileSystemCreated(config *osdConfig) error {
+	if config.partitionScheme == nil {
+		return nil
+	}
+
+	savedScheme, err := LoadScheme(config.kv, config.storeName)
+	if err != nil {
+		return fmt.Errorf("failed to load the saved partition scheme: %+v", err)
+	}
+
+	// mark the OSD's filesystem as created
+	config.partitionScheme.FSCreated = true
+	if err := savedScheme.updateSchemeEntry(config.partitionScheme); err != nil {
+		return fmt.Errorf("failed to update partition scheme entry %+v", config.partitionScheme)
+	}
+
+	if err := savedScheme.SaveScheme(config.kv, config.storeName); err != nil {
+		return fmt.Errorf("failed to save partition scheme: %+v", err)
+	}
+
+	return nil
+}
+
+func backupOSFileSystem(config *osdConfig, clusterName string) error {
+	if !isBluestore(config) {
+		return nil
+	}
+
+	logger.Infof("Backing up OSD %d file system from %s", config.id, config.rootPath)
+
+	storeName := fmt.Sprintf(osdFSStoreNameFmt, config.id)
+
+	// ensure the store we are backing up to is clear first
+	if err := config.kv.ClearStore(storeName); err != nil {
+		return err
+	}
+
+	fis, err := ioutil.ReadDir(config.rootPath)
+	if err != nil {
+		return err
+	}
+
+	filter := util.CreateSet([]string{
+		filepath.Base(mon.GetConfFilePath(config.rootPath, clusterName)), // filter out the rook.config file since it's always regenerated
```

Comment:
i think we should also filter out the keyring file.  we can regenerate it as needed and backing it up to a config map means we are putting a secret into a non secret resource.